analyzers/src/SonarAnalyzer.CSharp/Rules/PrivateStaticMethodUsedOnlyByNestedClass.cs

```diff
@@ -0,0 +1,190 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.CFG.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class PrivateStaticMethodUsedOnlyByNestedClass : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S3398";
+    private const string MessageFormat = "Move the method inside '{0}'.";
+
+    private static readonly SyntaxKind[] AnalyzedSyntaxKinds = new[]
+    {
+        SyntaxKind.ClassDeclaration,
+        SyntaxKind.StructDeclaration,
+        SyntaxKind.InterfaceDeclaration,
+        SyntaxKindEx.RecordClassDeclaration,
+        SyntaxKindEx.RecordStructDeclaration
+    };
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var declaredType = (TypeDeclarationSyntax)c.Node;
+
+                if (!IsPartial(declaredType)
+                    && HasNestedTypeDeclarations(declaredType)
+                    && PrivateStaticMethodsOf(declaredType) is { Length: > 0 } candidates)
+                {
+                    var methodReferences = TypesWhichUseTheMethods(candidates, declaredType, c.SemanticModel);
+
+                    foreach (var reference in methodReferences)
+                    {
+                        var typeToMoveInto = LowestCommonAncestorOrSelf(reference.Types);
+                        if (typeToMoveInto != declaredType)
+                        {
+                            var nestedTypeName = typeToMoveInto.Identifier.ValueText;
+                            c.ReportIssue(Diagnostic.Create(Rule, reference.Method.Identifier.GetLocation(), nestedTypeName));
+                        }
+                    }
+                }
+            },
+            AnalyzedSyntaxKinds);
+
+    private static MethodDeclarationSyntax[] PrivateStaticMethodsOf(TypeDeclarationSyntax type) =>
```

Comment:
nitpick: static private methods at the end

Reply:
Every method in the class is private and static - except for `Initialize` - so technically they are at the end of the class (right before the nested type declarations). Did I miss something?

Reply:
my bad